### ISSUE TITLE
pixel weight

### ISSUE THREAD

**Issue by elfpattern** [open-mmlab/mmsegmentation#507]
Now, based on CrossEntropyLoss,   I want to add a pixel weight, but, how to add?   The forward function has the param, but no use！
![image](https://user-images.githubusercontent.com/26590396/115687837-261bbb80-a38d-11eb-8a9d-0d899571e21e.png)

**Comment by MengzhangLI** [comment]
Hi, sorry for late reply.

I suggest you use OHEM method, please refer to [training_tricks_OHEM](https://github.com/open-mmlab/mmsegmentation/blob/master/docs/tutorials/training_tricks.md#online-hard-example-mining-ohem) for detailed information. Default setting is truncating top 100,000 pixels which have high loss. 

Note: 
(1) OHEM class would use prediction logits output and its counterpart ground truth value. 
(2) You could also add `weight` in your config file just at below place:

![image](https://user-images.githubusercontent.com/24582831/123222897-fa3ec280-d502-11eb-9676-90d9a9a5fa5e.png)

**Comment by elfpattern** [comment]
Ok, thx, I have done the pixel weights!




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年6月24日(星期四) 下午3:46
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [open-mmlab/mmsegmentation] pixel weight (#507)





 
Hi, sorry for late reply.
 
I suggest you use OHEM method, please refer to training_tricks_OHEM for detailed information. Default setting is truncating top 100,000 pixels which have high loss.
 
Note:
 (1) OHEM class would use prediction logits output and its counterpart ground truth value.
 (2) You could also add weight in your config file just at below place:

**Comment by MengzhangLI** [comment]
Could you list your method if it is different from what I suggest?

I guess your method may be helpful for other users of MMSegmentation.

Best,

**Comment by elfpattern** [comment]
Happy to share, First my task is two classes,&nbsp; I have two solutions, one to change the dataloader, which add a edge pixel indexes. the other is to add a class which name is base_pixel_sample, but this solution needs to convert tensor to numpy , then to tensor, Fortunately, It is not time consuming.
mport torch
import torch.nn.functional as F

from ..builder import PIXEL_SAMPLERS
from .base_pixel_sampler import BasePixelSampler
import cv2
import numpy as np

kernel = np.ones((5, 5), np.uint8)


@PIXEL_SAMPLERS.register_module()
class EdgePixelSampler(BasePixelSampler):
    """Online Hard Example Mining Sampler for segmentation.

    Args:
        context (nn.Module): The context of sampler, subclass of
            :obj:`BaseDecodeHead`.
    """

    def __init__(self, context):
        super(EdgePixelSampler, self).__init__()
        self.context = context

    def get_edge_index(self, seg_label):
        global kernel
        seg_label_numpy = seg_label.cpu().data.numpy()
        edge_index = np.zeros(seg_label_numpy.shape)

        for i in range(seg_label_numpy.shape[0]):
            one_seg = seg_label_numpy[i]
            one_seg = one_seg.astype('uint8')
            dilate = cv2.dilate(one_seg, kernel)
            erode = cv2.erode(one_seg, kernel)
            final = dilate - erode
            edge_index[i, ...] = np.where(final != 0, 1, edge_index[i, ...])
        return edge_index

    def sample(self, seg_logit, seg_label):
        """Sample pixels that have high loss or with low prediction confidence.

        Args:
            seg_logit (torch.Tensor): segmentation logits, shape (N, C, H, W)
            seg_label (torch.Tensor): segmentation label, shape (N, 1, H, W)

        Returns:
            torch.Tensor: segmentation weight, shape (N, H, W)
        """

        with torch.no_grad():
            assert seg_logit.shape[2:] == seg_label.shape[2:]
            assert seg_label.shape[1] == 1
            # 去掉维度为1 的那一维，是(N, H, W)
            seg_label = seg_label.squeeze(1).long()
            valid_mask = seg_label != self.context.ignore_index
            # seg_weight的维度是(N, H, W)
            seg_weight = seg_logit.new_ones(size=seg_label.size())
            ##处理seg_label, 是(N, H, W)
            edge_index = self.get_edge_index(seg_label)
            edge_index_tensor = torch.from_numpy(edge_index)
            seg_weight[edge_index_tensor != 0] = 2
            return seg_weight





------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2021年6月24日(星期四) 晚上7:25
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [open-mmlab/mmsegmentation] pixel weight (#507)





 
Could you list your method if it is different from what I suggest?
 
I guess your method may be helpful for other users of MMSegmentation.
 
Best,

**Comment by MengzhangLI** [comment]
Got it. Thanks!

Best,
